cranelift/codegen/src/isa/aarch64/lower.isle

```diff
@@ -2637,16 +2637,18 @@
            (y_hi Reg (value_regs_get y_regs 1)))
         ;; cannot use the with_flags helper here but it should be fine right now
         (let    
-            ((lo Reg (alu_rrr alu_op1 $I64 x_lo y_lo))
-             (hi Reg (alu_rrr alu_op2 $I64 x_hi y_hi))
-             (of WritableReg (let 
-                                    ((dst WritableReg (temp_writable_reg $I64))
-                                    (_ Unit (emit (MInst.CSet dst cond))))
-                              dst)
-            ))
+            ((lo_inst ProducesFlags (alu_rrr_with_flags_paired $I64 x_lo y_lo alu_op1))
+             (hi_inst ProducesFlags (alu_rrr_with_flags_paired $I64 x_hi y_hi alu_op2))
+             (of_inst ConsumesFlags (cset_paired cond))
+             
+             (result InstOutputRef (output_to_output_ref (with_flags_extended (produces_flags_concat lo_inst hi_inst) of_inst)))
```

Comment:
Hmm, now that I see this with `InstOutputRef`, I don't think this is very readable either; making the "extended" variant of the combinator take *only* the specific combination we need feels odd too. The second `ProducesFlags` also consumes flags from the previous op, which makes it a bit of an impedance mismatch too.

I think something like this would (finally!) work though:

```
(type MultiReg (Three (a Reg) (b Reg) (c Reg))

(decl with_flags_chain (ProducesFlags ProducesAndConsumesFlags ConsumesFlags) MultiReg)

;; inst constructor that turns `adcs` (aarch64) or `add` (x64) into a `ProducesAndConsumesFlags`

;; then:

(decl pair_and_single (MultiReg) InstOutput)
(rule (pair_and_single (MultiReg.Three a b c)
  (output_pair (value_regs a b) c))
```

does that seem reasonable?

Reply:
> The second `ProducesFlags` also consumes flags from the previous op, which makes it a bit of an impedance mismatch too.
That is true, though the `ccmp` constructor on arm does the same thing with `ProducesFlagsTwiceSideEffect`.

The rest is reasonable, i guess, though I'd need to implement more generic inst constructors for `alu_rmi_r`/`alu_rrr` that construct a `ConsumesAndProducesFlags` (probably better ordered like that) since the helpers do not know which alu op they are constructing specifically.

Reply:
I replaced the old change with what you proposed. It supports more combinations now, though that ended up in a lot of copypaste.